src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -190,81 +190,126 @@ protected virtual string GenerateUniqueId()
-                    kvp => string.Equals(kvp.Key, attributeDescriptor.Name, StringComparison.OrdinalIgnoreCase));
+                var attributeName = attributeKeyValuePair.Key;
+                var attributeValueChunk = attributeKeyValuePair.Value;
+                if (attributeValueChunk == null)
+                {
+                    // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
+                    // logged an error if it was a bound attribute; so we can skip.
+                    continue;
+                }
 
-                if (matchingAttributes.Any())
+                // Find the matching TagHelperAttributeDescriptor.
+                var attributeDescriptor = attributeDescriptors.FirstOrDefault(
+                    descriptor => descriptor.IsNameMatch(attributeName));
+                if (attributeDescriptor == null)
                 {
-                    // First attribute wins, even if there's duplicates.
-                    var firstAttribute = matchingAttributes.First();
-                    var attributeValueChunk = firstAttribute.Value;
+                    // Attribute is not bound to a property or indexer in this tag helper.
+                    continue;
+                }
 
-                    // Minimized attributes are not valid for bound attributes. There will be an error for the bound
-                    // attribute logged by TagHelperBlockRewriter already so we can skip.
-                    if (attributeValueChunk == null)
+                // We capture the tag helper's property value accessor so we can retrieve it later (if we need to).
+                var valueAccessor = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "{0}.{1}",
+                    tagHelperVariableName,
+                    attributeDescriptor.PropertyName);
+
+                if (attributeDescriptor.IsIndexer)
+                {
+                    // Need a different valueAccessor in this case. But first need to throw a reasonable Exception at
+                    // runtime if the property is null. The check is not required at design time.
+                    if (!_designTimeMode && confirmedDictionaries.Add(attributeDescriptor.PropertyName))
                     {
-                        continue;
+                        _writer
+                            .Write("if (")
+                            .Write(valueAccessor)
+                            .WriteLine(" == null)");
+                        using (_writer.BuildScope())
+                        {
+                            // System is in Host.NamespaceImports for all MVC scenarios. No need to generate FullName
+                            // of InvalidOperationException type.
+                            _writer
+                                .Write("throw ")
+                                .WriteStartNewObject(typeof(InvalidOperationException).Name)
```

Comment:
That was a nit, right? I'll do it but could just as well pass `"InvalidOperationException"` since the name can't be refactored to anything else.